Document/0x07a-Testing-Authentication-and-Session-Management.md

```diff
@@ -66,6 +66,106 @@ If any of these two conditions raise an issue, reject the request and do not all
 
 
 
+### Testing JSON Web Token (JWT)
+
+#### Overview
+
+JSON Web Token (JWT) ensures the integrity of information within a JSON object between two parties and is defined in RFC 7519<sup>[1]</sup>. A cryptographic signature is created for the data within the token. This only allows the server to create and modify tokens and enables a stateless authentication. The server doesn't need to remember any session or any other authentication information, as everything is contained within JWT.
+
+An example of an encoded JSON Web Token can be found below<sup>[5]</sup>.
+
+```
+eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9.TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ
+```
+
+JWTs are Base-64 encoded and are divided into three parts:
+
+* **Header** Algorithm and Token Type (eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9):
+```JSON
+{"alg":"HS256","typ":"JWT"}
+```
+* **Claims** Data  (eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9):
+```JSON
+{"sub":"1234567890","name":"John Doe","admin":true}
+```
+* **JSON Web Signature (JWS)** (TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ):
+```JSON
+HMACSHA256(
+  base64UrlEncode(header) + "." +
+  base64UrlEncode(payload),
+  secret
+)
+```
+
+For mobile apps it's more and more used to authenticate both the message sender and receiver by using JWT. JWT implementations are available for all major programming languages, like PHP<sup>[2]</sup> or Java Spring<sup>[3]</sup>.
+
+#### Static Analysis
+
+Identify the JWT library that is used on server and client side. Check if there are any known vulnerabilities available for the JWT libraries in use.
+
+The following best practices should be verified in the JWT libraries<sup>[7]</sup>:
+* Verify the signature on server-side at all times for all incoming requests containing a token.
+* Verify the data embedded into the JWT and check if sensitive information is available. If so, it should be considered to encrypt the JWT content.
+* Verify if replay attacks are addressed by using `jti` (JWT ID) claim, which provides a unique identifier for JWT.
+* Verify where the secret signing key is located and stored. It should only be available for the issuer and consumer.
```

Comment:
it's either the private signing key or the secret HMACing key. I would say that it is only available to the issuer and the verifier. Because saying consumer might mean that others believe they can use it too.